### PR TITLE
feat(melange): implement promotion for melange.emit

### PR DESCRIPTION
**Pull request by anmonteiro**
Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

### CODE REVIEW

File: src/dune_rules/melange_rules.ml [hunk 1]
```diff
@@ -43,7 +43,7 @@ let js_includes ~sctx ~target_dir ~requires_link ~scope ~js_ext =
              ; Hidden_deps hidden_deps
              ])))
 
-let build_js ~loc ~dir ~pkg_name ~module_system ~dst_dir ~obj_dir ~sctx
+let build_js ~loc ~dir ~pkg_name ~mode ~module_system ~dst_dir ~obj_dir ~sctx
```
Comment (jchavarri) [inline src/dune_rules/melange_rules.ml:46]: Maybe out of scope, but there are already 4 params that depend on `mel` value: `loc`, `mode`, `module_system` and `js_ext`. I wonder at what point it's just easier to pass the `mel` value directly.

File: src/dune_rules/melange_rules.ml [hunk 2]
```diff
@@ -78,7 +78,7 @@ let build_js ~loc ~dir ~pkg_name ~module_system ~dst_dir ~obj_dir ~sctx
        ])
 
 let add_rules_for_entries ~sctx ~dir ~expander ~dir_contents ~scope
-    ~compile_info ~target_dir (mel : Melange_stanzas.Emit.t) =
+    ~compile_info ~target_dir ~mode (mel : Melange_stanzas.Emit.t) =
```
Comment (jchavarri) [inline src/dune_rules/melange_rules.ml:81]: If `Rule.Mode.t` was derived just before calling `Super_context.add_rule`, we'd avoid the param here.

**Review by jchavarri — APPROVED** [review]
Looks great 😍 Some refactor ideas, but all optional.

File: src/dune_rules/melange_rules.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/melange_rules.ml:46]: Even though it's easier to pass more than what we need, we pay for that by making it harder to understand the function.

File: test/blackbox-tests/test-cases/melange/promote.t [hunk 3]
```diff
@@ -0,0 +1,35 @@
+Test melange.emit promotion
```
Comment (jchavarri) [inline test/blackbox-tests/test-cases/melange/promote.t:1]: Would it make sense to add a case with a library? As it follows a bit different logic than melange.emit.

File: src/dune_rules/melange_rules.ml (hunk 2 above)
Comment (anmonteiro) [reply src/dune_rules/melange_rules.ml:81]: I'd rather compute the mode once than avoid passing an argument.

File: test/blackbox-tests/test-cases/melange/promote.t (hunk 3 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/promote.t:1]: added in [540d021](https://github.com/ocaml/dune/pull/6500/commits/540d021f6efda0ae3633511c9ca12212960fd00d)

**Pull request merged**